File: 0110-Balanced Binary Tree/0110-Balanced Binary Tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    bool isBalanced(TreeNode* root) {
        int h = 0;
        return helper(root, h);
    }
    
private:
    bool helper(TreeNode* root, int& h) {
        if (root == nullptr) {
            return true;
        }
        
        int lh = 0, rh = 0;
        bool isLeftBalanced = helper(root->left, lh);
        bool isRightBalanced = helper(root->right, rh);
        h = max(lh, rh) + 1;
        return isLeftBalanced && isRightBalanced && abs(lh - rh) <= 1;
    }
};

File: 0101-0200/0110-Balanced Binary Tree/0110-Balanced Binary Tree.py
# Definition for a binary tree node.
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

class Solution:
    def isBalanced(self, root: TreeNode) -> bool:
        def dfs(root):
            if root is None:
                return (True, 0)
            lb, lh = dfs(root.left)
            rb, rh = dfs(root.right)
            h = max(lh, rh) + 1
            return (lb and rb and abs(lh - rh) <= 1, h)
        return dfs(root)[0]
